Add a `gr.Draggable` layout

abidlabs opened the pull request.

freddyaboulton reviewed (commented):
Looks great. Will continue to take a look tomorrow! One thing I noticed is that it doesn't play well with components that have a dragging mechanic like slider (maybe ImageEditor as well). Should we add a toggle to enable the layout dragging?

![cant_slide](https://github.com/user-attachments/assets/1fc03bf4-7bbd-4f74-a137-3cf889bd80e9)

abidlabs commented:
> I think this is causing the drag preview to always go to the bottom of the page

removed

> This may be adding a lot of duplicate event listeners since it's called on every mutation

good call, I removed the older event listeners. not sure if there's a better pattern here.

> The use of global may leak to other parts of the app or cause unintended consequences, no?

used more targeted css by adding a class to the draggable's children. we can't avoid global() because we are adding classes dynamically, but this should be safer.

abidlabs commented:
Ok everything should be resolved now @freddyaboulton and good for another pass!

pngwn commented:
What is the use case for this component? I haven't really seen this requested by anyone and I'm not sure how useful it is in general.

abidlabs commented:
> What is the use case for this component? I haven't really seen this requested by anyone and I'm not sure how useful it is in general.

The [immdiate motivation was for Trackio](https://huggingface.slack.com/archives/C02SPHC1KD1/p1755039182094199), but I think this could be quite useful (and a nice flashy feature) for dashboards more generally. As it stands, dashboards are not quite a common use case, but my feeling is that we can grow this use case a bit by showcasing Trackio as an example what you can build with Gradio.

pngwn commented:
If it was sticky for that user (persisted after reload) then I could see the utility but as it is i struggle to see how it is useful even for trackio. I'll test this properly tomorrow tho to see how it plays with other components.

pngwn reviewed (approved):
Looks great from my testing. Thanks @abidlabs!

abidlabs commented:
Thanks @freddyaboulton @dawoodkhan82 @pngwn for the detailed reviews!
